clang/lib/Parse/ParseExprCXX.cpp

```diff
@@ -1576,7 +1576,10 @@ ExprResult Parser::ParseLambdaExpressionAfterIntroducer(
                       TrailingReturnTypeLoc, &DS),
                   std::move(Attributes), DeclEndLoc);
 
-    Actions.ActOnLambdaClosureQualifiers(Intro, MutableLoc);
+    // We have called ActOnLambdaClosureQualifiers for parentheses-less cases
```

Comment:
Is there any good reason to do this in 2 places, rather than just delay the above one to here?  @cor3ntin ?

Reply:
Yeah, looking into `ActOnLambdaClosureQualifiers`, I don't see anything that (currently) depends on lambda parameters. I guess we can combine these two calls? I'd love to open a separate NFC PR for doing so, before landing this patch. :)

@erichkeane I retrospected your comment in https://github.com/llvm/llvm-project/pull/78598#discussion_r1478458713 and I appreciate the diagnostics for non-ODR uses - however that's much like a QoI issue to me and would it be OK to leave it as a follow-up? Let me know what you think and I'll add a FIXME then.

Reply:
I think I'd be ok with this as a followup.  But @cor3ntin is the one who knows the most about lambda stuff (particularly parsing, etc), so I'd like to see him approve this.

Reply:
Yes, we need to set whether the lambda has a `mutable` keyword as soon as the parameters, if present, are parsed, as that affect the meaning of subsequent id-expressions

When there are no parens, there can be a noexcept clause for example, and parsing that does requires knowing whether the lambda is mutable, so that we can adjust the type of any reference-to-capture.

Reply:
> Yes, we need to set whether the lambda has a mutable keyword as soon as the parameters, if present, are parsed, as that affect the meaning of subsequent id-expressions
> When there are no parens, there can be a noexcept clause for example, and parsing that does requires knowing whether the lambda is mutable, so that we can adjust the type of any reference-to-capture.

Thanks, that makes sense! We currently set `LSI->Mutable` in `ActOnLambdaClosureQualifiers()` , and while I think it's possible to move it out of the function, I'm not opposed to leaving it as-is.